src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -534,44 +526,88 @@ private static async Task WriteDataAsync(Stream archiveStream, Stream dataStream
             cancellationToken.ThrowIfCancellationRequested();
 
             await dataStream.CopyToAsync(archiveStream, cancellationToken).ConfigureAwait(false); // The data gets copied from the current position
+
             int paddingAfterData = TarHelpers.CalculatePadding(actualLength);
-            await archiveStream.WriteAsync(new byte[paddingAfterData], cancellationToken).ConfigureAwait(false);
+            if (paddingAfterData != 0)
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(paddingAfterData);
+                Array.Clear(buffer, 0, paddingAfterData);
+
+                await archiveStream.WriteAsync(buffer.AsMemory(0, paddingAfterData), cancellationToken).ConfigureAwait(false);
+
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
         }
 
         // Dumps into the archive stream an extended attribute entry containing metadata of the entry it precedes.
         private static Stream? GenerateExtendedAttributesDataStream(Dictionary<string, string> extendedAttributes)
         {
             MemoryStream? dataStream = null;
+
+            byte[]? buffer = null;
+            Span<byte> span = stackalloc byte[512];
+
             if (extendedAttributes.Count > 0)
             {
                 dataStream = new MemoryStream();
+
                 foreach ((string attribute, string value) in extendedAttributes)
                 {
-                    byte[] entryBytes = GenerateExtendedAttributeKeyValuePairAsByteArray(Encoding.UTF8.GetBytes(attribute), Encoding.UTF8.GetBytes(value));
-                    dataStream.Write(entryBytes);
+                    // Generates an extended attribute key value pair string saved into a byte array, following the ISO/IEC 10646-1:2000 standard UTF-8 encoding format.
+                    // https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html
+
+                    // The format is:
+                    //     "XX attribute=value\n"
+                    // where "XX" is the number of characters in the entry, including those required for the count itself.
+                    int length = 3 + Encoding.UTF8.GetByteCount(attribute) + Encoding.UTF8.GetByteCount(value);
+                    length += CountDigits(length);
+
+                    // Get a large enough buffer if we don't already have one.
+                    if (span.Length < length)
+                    {
+                        if (buffer is not null)
```

Comment:
when is buffer not null here?

Reply:
When we previously rented one.  We're in a loop.